### PR TITLE
[execution] add epoch stable sequence number to transaction effects for configs

### PR DESCRIPTION
**Pull request by tzakian**
## Description 

This adds support for including the epoch-stable sequence number for config objects (and the deny list object) in transaction effects. This is useful in general, but particularly so for transaction replay. 

This is done by adding a new marker value to the marker table that holds the sequence number at which the object was mutated for any config mutations or mutations to the deny list object. Importantly, there is only one marker placed per-object per-epoch (this is ensured by  always placing the value at version 0, and query for the existence of the key before placing the marker in the authority store). When we look for the epoch stable sequence number for the object, we either return the value in the marker table (if present) or the current version of the object. 

This deprecates the old `UnchangedSharedKind::PerEpochConfig` and replaces this with `UnchangedSharedKind::PerEpochConfigWithSeqno(SequenceNumber)`. 

This adds a new `ConfigStore` trait that enables accessing the per-epoch stable sequence number for a given object. 

This additionally fixes an error in `reconfigure_for_testing` where we were not properly initializing randomness after a reconfiguration in test scenarios. 

## Test plan 

Added a unit test that checks that with different transaction orderings (and mutations) the epoch-stable sequence numbers are fact stable. Additionally updated the e2e `per_epoch_config_stress_test` to make sure epoch-stable sequence numbers are stable as well.

---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates. 

- [X] Protocol: Added epoch-stable sequence numbers for read-only per-epoch configs accessed in the transaction. 
- [ ] Nodes (Validators and Full nodes): 
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [ ] CLI: 
- [ ] Rust SDK:

### CODE REVIEW

File: sui-execution/latest/sui-adapter/src/temporary_store.rs [hunk 1]
```diff
@@ -199,6 +204,64 @@ impl<'backing> TemporaryStore<'backing> {
...
+                }
+            }
+        }
+
+        for id in loaded_per_epoch_config_objects.iter() {
+            let seqno = self
+                .store
+                .get_current_epoch_stable_sequence_number(id, self.cur_epoch)
+                .expect("Should panic on storage error")
+                .expect(
```
Comment (tzakian) [inline sui-execution/latest/sui-adapter/src/temporary_store.rs:250]: Note the `expect`s here. These are safe since:
1. We should panic on a storage error and not raise any other type of error. 
2. The config (or deny list) can never be deleted, and therefore the object must always exist.

Comment (lxfind) [reply sui-execution/latest/sui-adapter/src/temporary_store.rs:250]: These comments should be in the code instead of as github comment

File: sui-execution/latest/sui-adapter/src/temporary_store.rs [hunk 2]
```diff
@@ -199,6 +204,64 @@ impl<'backing> TemporaryStore<'backing> {
...
+                    "Config object already loaded during execution. Must be able to be reloaded.",
+                );
+            if let Some(other_seqno) = loaded_configs_at_version.insert(*id, Some(seqno)) {
```
Comment (lxfind) [inline sui-execution/latest/sui-adapter/src/temporary_store.rs:253]: If you simply map `loaded_per_epoch_config_objects`, then you probably don't need this check since it's a BTreeSet.

File: sui-execution/latest/sui-adapter/src/temporary_store.rs [hunk 3]
```diff
@@ -199,6 +204,64 @@ impl<'backing> TemporaryStore<'backing> {
...
+                BTreeMap::new(),
+            );
+        }
+
+        let mut mutated_configs_at_version = BTreeMap::new();
+        let mut loaded_configs_at_version = BTreeMap::new();
+        for (id, obj) in self.input_objects.iter() {
+            if obj.struct_tag().is_some_and(is_config) || id == &SUI_DENY_LIST_OBJECT_ID {
+                if let Some(((seqno, _), _)) = self.mutable_input_refs.get(id) {
+                    mutated_configs_at_version.insert(*id, *seqno);
```
Comment (lxfind) [inline sui-execution/latest/sui-adapter/src/temporary_store.rs:240]: Use filter_map instead?

File: crates/sui-types/src/inner_temporary_store.rs [hunk 4]
```diff
@@ -34,6 +34,8 @@ pub struct InnerTemporaryStore {
     pub binary_config: BinaryConfig,
     pub runtime_packages_loaded_from_db: BTreeMap<ObjectID, PackageObject>,
     pub lamport_version: SequenceNumber,
+    // Mapping of the mutated config objects to their new sequence numbers at the time of mutation.
+    pub mutated_config_objects: BTreeMap<ObjectID, SequenceNumber>,
```
Comment (lxfind) [inline crates/sui-types/src/inner_temporary_store.rs:38]: Do you have to pass it through from the temp store? Could you compute it in the transaction output instead?

File: crates/sui-core/src/authority/authority_store_tables.rs [hunk 5]
```diff
@@ -345,6 +345,22 @@ impl AuthorityPerpetualTables {
...
     }
 
+    pub fn get_current_epoch_stable_sequence_number(
+        &self,
+        object_id: &ObjectID,
+        epoch_id: EpochId,
+    ) -> SuiResult<Option<VersionNumber>> {
+        let object_key = (epoch_id, ObjectKey::min_for_id(object_id));
+        let object = self.get_object(object_id);
+        Ok(match self.object_per_epoch_marker_table.get(&object_key)? {
```
Comment (lxfind) [inline crates/sui-core/src/authority/authority_store_tables.rs:355]: Can you add a comment why the two DB reads (get_object and get marker) are data-race safe?

File: crates/sui-core/src/authority/authority_store.rs [hunk 6]
```diff
@@ -883,10 +883,28 @@ impl AuthorityStore {
         // Add batched writes for objects and locks.
         let effects_digest = effects.digest();
 
+        // Get the already-written markers for configs this epoch.
+        let mut already_written_config_updates = self
+            .perpetual_tables
+            .object_per_epoch_marker_table
+            .multi_contains_keys(
```
Comment (lxfind) [inline crates/sui-core/src/authority/authority_store.rs:890]: Can you add a comment why this read is data-race free?

File: crates/sui-core/src/authority/authority_per_epoch_store.rs [hunk 7]
```diff
@@ -1040,27 +1040,53 @@ impl AuthorityPerEpochStore {
...
-            backing_package_store,
-            object_store,
-            expensive_safety_check_config,
+            authority.get_backing_package_store().clone(),
+            authority.get_object_store().clone(),
+            &authority.config.expensive_safety_check_config,
             self.chain_identifier,
+        );
+
+        // Set up randomness manager for the next epoch if randomness is enabled.
```
Comment (tnowacki) [inline crates/sui-core/src/authority/authority_per_epoch_store.rs:1067]: I'm a bit confused at the nature of these changes. Is it a bad merge or some refactoring?

File: crates/sui-core/src/authority/authority_store.rs [hunk 8]
```diff
@@ -883,10 +883,28 @@ impl AuthorityStore {
...
+
         write_batch.insert_batch(
             &self.perpetual_tables.object_per_epoch_marker_table,
             markers
                 .iter()
+                .filter(|(_, marker_value)| {
+                    !matches!(marker_value, MarkerValue::ConfigUpdate(_))
+                        || !already_written_config_updates
+                            .pop()
+                            .expect("Information on all config updates are present by `multi_contains_keys` above")
```
Comment (tnowacki) [inline crates/sui-core/src/authority/authority_store.rs:906]: Are we just asserting that these two vectors line up properly? 
Should we just use a map instead?

File: crates/sui-core/src/execution_cache.rs [hunk 9]
```diff
@@ -455,6 +474,17 @@ pub trait ObjectCacheRead: Send + Sync {
...
 
+    fn have_updated_config_in_epoch(
+        &self,
+        object_id: &ObjectID,
+        epoch_id: EpochId,
+    ) -> SuiResult<bool> {
+        match self.get_marker_value(object_id, SequenceNumber::new(), epoch_id)? {
+            Some(MarkerValue::ConfigUpdate(_)) => Ok(true),
+            _ => Ok(false),
+        }
```
Comment (tnowacki) [inline crates/sui-core/src/execution_cache.rs:485]: Should we just call `get_last_config_object_change_sequence_number` with `is_some()`?

File: crates/sui-core/src/unit_tests/data/coin_deny_list_v2/sources/regulated_coin.move [hunk 10]
```diff
@@ -17,6 +17,8 @@ module coin_deny_list_v2::regulated_coin {
             true,
             ctx
         );
+        let coin = coin::mint(&mut treasury_cap, 1000000, ctx);
+        transfer::public_transfer(coin, ctx.sender());
```
Comment (tnowacki) [inline crates/sui-core/src/unit_tests/data/coin_deny_list_v2/sources/regulated_coin.move:21]: What is this change?

File: crates/sui-e2e-tests/tests/per_epoch_config_stress_tests.rs [hunk 11]
```diff
@@ -63,21 +69,46 @@ async fn run_thread<F, Fut>(
...
             continue;
         };
+
+        // make sure we have the raw effects
+        assert!(!raw_effects.is_empty());
+
+        let Ok(TransactionEffects::V2(v2_effects)) =
+            bcs::from_bytes::<TransactionEffects>(&raw_effects)
+        else {
+            panic!("Failed to deserialize raw effects into effects v2");
```
Comment (tnowacki) [inline crates/sui-e2e-tests/tests/per_epoch_config_stress_tests.rs:93]: `.expect`

File: crates/sui-types/src/config.rs [hunk 12]
```diff
@@ -87,6 +87,20 @@ pub fn is_setting(tag: &StructTag) -> bool {
         && type_params.len() == 1
 }
 
+/// Returns true if the given `tag` is a `Config` struct.
+pub fn is_config(tag: StructTag) -> bool {
```
Comment (tnowacki) [inline crates/sui-types/src/config.rs:91]:
```suggestion
pub fn is_config(tag: &StructTag) -> bool {
```

File: crates/sui-types/src/inner_temporary_store.rs (hunk 4 above)
Comment (tnowacki) [reply crates/sui-types/src/inner_temporary_store.rs:38]: +1

**Review by tnowacki — COMMENTED** [review]
Here are some initial comments, I have more reading to do though

File: sui-execution/latest/sui-adapter/src/temporary_store.rs [hunk 13]
```diff
@@ -199,6 +204,64 @@ impl<'backing> TemporaryStore<'backing> {
...
+                    .map(|id| (*id, None))
+                    .collect(),
+                BTreeMap::new(),
+            );
+        }
+
+        let mut mutated_configs_at_version = BTreeMap::new();
+        let mut loaded_configs_at_version = BTreeMap::new();
+        for (id, obj) in self.input_objects.iter() {
+            if obj.struct_tag().is_some_and(is_config) || id == &SUI_DENY_LIST_OBJECT_ID {
```
[the remark this answers is not in the document]
Comment (tzakian) [reply sui-execution/latest/sui-adapter/src/temporary_store.rs:238]: Not sure I follow here. We currently only support `0x403` however this supports arbitrary configs + the `0x403` object. So we're saying here "if it's a config or if it's the 0x403 object" track it as a mutated config object.

File: crates/sui-types/src/inner_temporary_store.rs (hunk 4 above)
Comment (tzakian) [reply crates/sui-types/src/inner_temporary_store.rs:38]: I initially started out with it in transaction output computation however it felt a bit odd to me to do it there (with storage reading and the like). It felt a bit better here and I like having the computation for both the loaded config objects and mutated config objects in one place (loaded config object stuff will need to stay in `into_effects` due to them needing to be passed to effects generation). 

That being said I don't have a super strong opinion on it, so can move it y'all would prefer.

File: crates/sui-core/src/authority/authority_per_epoch_store.rs (hunk 7 above)
Comment (tzakian) [reply crates/sui-core/src/authority/authority_per_epoch_store.rs:1067]: Bugfix -- the `reconfigure_for_testing` function (which calls this) would fail any transaction after reconfiguration since it would not include the randomness manager (even if the previous epoch did). This fixes this by setting up the randomness manager for the next epoch if the previous epoch had randomness enabled.

File: crates/sui-core/src/authority/authority_store.rs (hunk 8 above)
Comment (tzakian) [reply crates/sui-core/src/authority/authority_store.rs:906]: Essentially, although a bit more complex -- we are asserting that the two vectors over the subset of `ConfigUpdate` marker values line up.

Note that the `markers` here are all markers to write (shared deleted, received, etc) and we don't want to not write those out or call `multi_contains_keys` for those so we only perform this query for config updates, and then as we see them we make sure they're in-synch. I might take a look at making this a bit nicer, as this is a bit hard to follow probably...

File: crates/sui-core/src/unit_tests/data/coin_deny_list_v2/sources/regulated_coin.move (hunk 10 above)
Comment (tzakian) [reply crates/sui-core/src/unit_tests/data/coin_deny_list_v2/sources/regulated_coin.move:21]: This is to support the new tests in `config_tests.rs`. So we can transfer regulated coins.

File: crates/sui-e2e-tests/tests/per_epoch_config_stress_tests.rs (hunk 11 above)
Comment (tzakian) [reply crates/sui-e2e-tests/tests/per_epoch_config_stress_tests.rs:93]: I was trying to collapse it all to one error (incorrect effects version, and invalid deser). Adding the `expect` might make it more complicated?

File: crates/sui-core/src/authority/authority_per_epoch_store.rs (hunk 7 above)
Comment (tnowacki) [reply crates/sui-core/src/authority/authority_per_epoch_store.rs:1067]: Might have been good as a separate PR, I might have been less confused then, but fine as is

File: crates/sui-core/src/authority/authority_store.rs [hunk 14]
```diff
@@ -883,10 +883,40 @@ impl AuthorityStore {
...
+            .object_per_epoch_marker_table
+            .multi_contains_keys(config_update_markers.iter().map(|(k, _)| (epoch_id, *k)))?;
+        assert_eq!(
+            already_written_config_updates.len(),
+            config_update_markers.len(),
+            "Already written config update information should have the same length as the config updates"
+        );
+        let config_update_markers_to_write = config_update_markers
+            .into_iter()
+            .zip(already_written_config_updates)
```
Comment (tnowacki) [inline crates/sui-core/src/authority/authority_store.rs:906]: A zip is definitely easier to follow, thanks!

File: crates/sui-core/src/authority/authority_store.rs [hunk 15]
```diff
@@ -883,10 +883,40 @@ impl AuthorityStore {
...
+            .filter_map(|(marker_key, already_written)| {
+                if !already_written {
+                    Some(marker_key)
+                } else {
+                    None
+                }
```
Comment (tnowacki) [inline crates/sui-core/src/authority/authority_store.rs:912]: For your consideration: a `filter` + a `map` would be less lines of code

File: crates/sui-core/src/transaction_outputs.rs [hunk 16]
```diff
@@ -74,6 +76,28 @@ impl TransactionOutputs {
...
+            // Create markers for each config update. Note that we always place it under the same
+            // object key. This way we can easily query to see if the config marker has already
+            // been updated this epoch.
+            let mutated_config_objects = input_objects
+                .iter()
+                .filter_map(|(id, obj)| {
+                    if obj.struct_tag().is_some_and(|tag| is_config(&tag))
+                        || *id == SUI_DENY_LIST_OBJECT_ID
+                    {
+                        if let Some(((seqno, _), _)) = mutable_inputs.get(id) {
```
Comment (tnowacki) [inline crates/sui-core/src/transaction_outputs.rs:88]: There was some RFC a while back in Rust for `let-and`, where you could put `if-let` in conditionals... really wishing we had that here

File: crates/sui-protocol-config/src/snapshots/sui_protocol_config__test__version_60.snap [hunk 17]
```diff
@@ -333,3 +334,4 @@ checkpoint_summary_version_specific_data: 1
 max_soft_bundle_size: 5
 bridge_should_try_to_finalize_committee: true
 max_accumulated_txn_cost_per_object_in_mysticeti_commit: 10
+
```
Comment (tnowacki) [inline crates/sui-protocol-config/src/snapshots/sui_protocol_config__test__version_60.snap:337]: I'm surprised it added newlines here... not an issue, it is just intriguing

File: crates/sui-types/src/effects/effects_v2.rs [hunk 18]
```diff
@@ -411,7 +412,7 @@ impl TransactionEffectsV2 {
         executed_epoch: EpochId,
         gas_used: GasCostSummary,
         shared_objects: Vec<SharedInput>,
-        loaded_per_epoch_config_objects: BTreeSet<ObjectID>,
+        loaded_per_epoch_config_objects: BTreeMap<ObjectID, Option<SequenceNumber>>,
```
Comment (tnowacki) [inline crates/sui-types/src/effects/effects_v2.rs:415]: I really really wish `Rust` had `Either` in the stdlib. 
Just as a small note, I feel like this should be `Either<BTreeSet<ObjectID>, BTreeMap<ObjectID, SequenceNumber>>`. I paused at first because I was confused under what situations we wouldn't have the number, but if I followed correctly, it is basically an `Either<(), SequenceNumber>` gated on the `ProtocolConfig`

At least maybe a comment or debug assert that they are all none or all some

File: sui-execution/latest/sui-adapter/src/programmable_transactions/execution.rs [hunk 19]
```diff
@@ -108,11 +108,15 @@ mod checked {
         // for expensive invariant checks
         let wrapped_object_containers = object_runtime.wrapped_object_containers();
 
+        // We record the config objects that we have accessed in this transaction.
```
Comment (tnowacki) [inline sui-execution/latest/sui-adapter/src/programmable_transactions/execution.rs:111]:
```suggestion
        // We record the config objects that we have accessed in this transaction through the unsequenced read API
```

File: sui-execution/latest/sui-adapter/src/temporary_store.rs [hunk 20]
```diff
@@ -1038,6 +1074,12 @@ impl<'backing> Storage for TemporaryStore<'backing> {
         }
         result
     }
+
+    fn save_accessed_config_objects(&mut self, accessed_config_objects: BTreeSet<ObjectID>) {
```
Comment (tnowacki) [inline sui-execution/latest/sui-adapter/src/temporary_store.rs:1078]:
```suggestion
    fn save_unsequenced_config_accesses(&mut self, accessed_config_objects: BTreeSet<ObjectID>) {
```

File: sui-execution/latest/sui-adapter/src/temporary_store.rs [hunk 21]
```diff
@@ -241,8 +277,8 @@ impl<'backing> TemporaryStore<'backing> {
         let object_changes = self.get_object_changes();
 
         let lamport_version = self.lamport_timestamp;
-        // TODO: Cleanup this clone. Potentially add unchanged_shraed_objects directly to InnerTempStore.
-        let loaded_per_epoch_config_objects = self.loaded_per_epoch_config_objects.read().clone();
+        let loaded_config_objects =
+            self.compute_config_accesses(&self.loaded_per_epoch_config_objects.read());
```
Comment (tnowacki) [inline sui-execution/latest/sui-adapter/src/temporary_store.rs:281]: Similarly we should make sure this is noted as just being unsequenced

File: sui-execution/latest/sui-adapter/src/temporary_store.rs [hunk 22]
```diff
@@ -199,6 +201,40 @@ impl<'backing> TemporaryStore<'backing> {
             .collect()
     }
 
+    // Compute the set of config objects and their epoch-stable sequence numbers that have been
+    // accessed in the transaction. 
+    fn compute_config_accesses(
```
Comment (tnowacki) [inline sui-execution/latest/sui-adapter/src/temporary_store.rs:206]:
```suggestion
    fn compute_unsequenced_config_accesses(
```

File: crates/sui-types/src/effects/effects_v2.rs (hunk 18 above)
Comment (tnowacki) [inline crates/sui-types/src/effects/effects_v2.rs:415]:
```suggestion
        unsequenced_per_epoch_config_objects: BTreeMap<ObjectID, Option<SequenceNumber>>,
```
I know this was the existing name, but we should really highlight the access pattern here

Comment (tzakian) [reply crates/sui-types/src/effects/effects_v2.rs:415]: Yep! Followed exactly. This is really an `Either<BTreeSet<ObjectID>, BTreeMap<ObjectID, SequenceNumber>>`. I think comment + a debug assert would make sense here -- will add that in :)

File: crates/sui-core/src/execution_cache/writeback_cache.rs [hunk 23]
```diff
@@ -608,12 +609,22 @@ impl WritebackCache {
...
-            .value_mut()
-            .insert(object_key.version(), marker_value);
+            .or_default();
+
+        // We always insert a marker for a config update in the marker table with the same sequence
+        // number (SequenceNumber::MIN) as the marker value, so that we can look up the first
+        // version at which a config is mutated in an epoch in the marker table using only the
+        // object ID and the epoch ID. Because of this we would violate the monoticity of the
+        // `CachedVersionMap` if we inserted more than once in an epoch.
+        if matches!(&marker_value, MarkerValue::ConfigUpdate(_)) && !cached_version_map.is_empty() {
```
Comment (tzakian) [inline crates/sui-core/src/execution_cache/writeback_cache.rs:623]: Note the comment here.

**Comment by tzakian** [comment]
Reviving this PR (after oh so very very long!)

File: crates/sui-core/src/transaction_outputs.rs (hunk 16 above)
Comment (mystenmark) [reply crates/sui-core/src/transaction_outputs.rs:160]: while this PR has been languishing, i think this actually landed, but we probably have to update our edition...

**Review by mystenmark — APPROVED** [review]
sui core changes look good

File: crates/sui-core/src/transaction_outputs.rs (hunk 16 above)
Comment (tzakian) [reply crates/sui-core/src/transaction_outputs.rs:160]: Yea, I went to update this exact spot because they added in 1.86, but then realized our edition is still on 1.85... :'( Something I may do later today/next week though as it's so nice to have
